com.unity.shadergraph/Editor/Data/Nodes/Input/Texture/CalculateLevelOfDetailTexture2DNode.cs

```diff
@@ -0,0 +1,120 @@
+using System.Linq;
+using UnityEngine;
+using UnityEditor.Graphing;
+using UnityEditor.ShaderGraph.Internal;
+using UnityEditor.ShaderGraph.Drawing.Controls;
+
+namespace UnityEditor.ShaderGraph
+{
+    [Title("Input", "Texture", "Calculate Level Of Detail Texture 2D")]
+    class CalculateLevelOfDetailTexture2DNode : AbstractMaterialNode, IGeneratesBodyCode, IMayRequireMeshUV
+    {
+        [SerializeField]
+        bool m_Clamp = true;
+
+        [ToggleControl]
+        public ToggleData clamp
+        {
+            get => new ToggleData(m_Clamp);
+            set
+            {
+                m_Clamp = value.isOn;
+                Dirty(ModificationScope.Node);
+            }
+        }
+
+        public const int OutputSlotLODId = 0;
+        public const int TextureInputId = 1;
+        public const int UVInput = 2;
+        public const int SamplerInput = 3;
+
+        const string kOutputSlotLODName = "LOD";
+        const string kTextureInputName = "Texture";
+        const string kUVInputName = "UV";
+        const string kSamplerInputName = "Sampler";
+
+        public override bool hasPreview { get { return true; } }
+
+        public CalculateLevelOfDetailTexture2DNode()
+        {
+            name = "Calculate Level Of Detail Texture 2D";
+            UpdateNodeAfterDeserialization();
+        }
+
+        public sealed override void UpdateNodeAfterDeserialization()
+        {
+            AddSlot(new Vector1MaterialSlot(OutputSlotLODId, kOutputSlotLODName, kOutputSlotLODName, SlotType.Output, 0, ShaderStageCapability.Fragment));
+            AddSlot(new Texture2DInputMaterialSlot(TextureInputId, kTextureInputName, kTextureInputName));
+            AddSlot(new UVMaterialSlot(UVInput, kUVInputName, kUVInputName, UVChannel.UV0));
+            AddSlot(new SamplerStateMaterialSlot(SamplerInput, kSamplerInputName, kSamplerInputName, SlotType.Input));
+
+            RemoveSlotsNameNotMatching(new[] { OutputSlotLODId, TextureInputId, UVInput, SamplerInput });
+        }
+
+        public override void Setup()
+        {
+            base.Setup();
+            var textureSlot = FindInputSlot<Texture2DInputMaterialSlot>(TextureInputId);
+            textureSlot.defaultType = Texture2DShaderProperty.DefaultType.White;
+        }
+
+        // Node generations
+        public virtual void GenerateNodeCode(ShaderStringBuilder sb, GenerationMode generationMode)
```

Comment:
Is there a better pattern for handling SHADER_TARGET versions? I used SampleTexture2DLODNode as my reference point.

Reply:
What I mean is a way to wrap up the entire ifdef stuff into a function. So something akin to writing this custom function node:
```
void CalculateLevelOfDetail_float(UnityTexture2D tex, float2 uv, UnitySamplerState s, out float lod)
{
    #if (SHADER_TARGET >= 41)
    lod = tex.CalculateLevelOfDetail(s, uv);
    #else
    lod = 0.5f*log2(max(dot(ddx(uv.xy * tex.texelSize.zw), ddx(uv.xy * tex.texelSize.zw)), dot(ddy(uv.xy * tex.texelSize.zw), ddy(uv.xy * tex.texelSize.zw))));
    lod = max(lod, 0);

    #if defined(MIP_COUNT_SUPPORTED)
        lod = min(lod, GetMipCount(tex.tex)-1);
    #endif
    
    #endif
}
```

Looking at this a bit more though, I'm actually not sure if we can do that due to samplers. So maybe it's fine as is. It might be doable using `TEXTURE2D_ARGS` but this might make things more complicated than it's worth, especially for now.

Reply:
I'll leave as-is, unless somebody has a better alternative, cheers.

Reply:
^ this should be 100% doable, to wrap it in a function.  You just have to make sure the function name is unique per variant of the function you are declaring.   i.e. m_Clamp and edgesSampler.Any() should change the name of the function.

This current inline form would work fine I think, it just results in more duplicated code when you have multiple instances of the node.